aquadoggo/src/graphql/client/dynamic_query.rs

```diff
@@ -0,0 +1,556 @@
+
+use std::convert::TryInto;
+
+use async_graphql::indexmap::IndexMap;
+use async_graphql::{
+    ContainerType, Context, Error, Name, SelectionField, ServerError, ServerResult, Value,
+};
+use async_recursion::async_recursion;
+use async_trait::async_trait;
+use futures::future;
+use log::{debug, error, info};
+use p2panda_rs::document::{DocumentId, DocumentView, DocumentViewId};
+use p2panda_rs::operation::OperationValue;
+use p2panda_rs::schema::SchemaId;
+
+use crate::db::provider::SqlStorage;
+use crate::db::traits::DocumentStore;
+use crate::graphql::client::dynamic_types::DocumentMetaType;
+use crate::graphql::client::utils::validate_view_matches_schema;
+use crate::graphql::scalars::{
+    DocumentId as DocumentIdScalar, DocumentViewId as DocumentViewIdScalar,
+};
+use crate::schema::SchemaProvider;
+
+/// Resolves queries for documents based on p2panda schemas.
+///
+/// Implements [`ContainerType`] to be able to resolve arbitrary fields selected by a query on the
+/// root GraphQL schema.
+///
+/// This implementation always has to match what is defined by the
+/// [`dynamic_output_type`][`crate::graphql::client::dynamic_types::dynamic_output_type`] module.
+#[derive(Debug, Default)]
+pub struct DynamicQuery;
+
+#[async_trait]
+impl ContainerType for DynamicQuery {
+    /// This resolver is called for all queries but we only want to resolve if the queried field
+    /// can actually be parsed in of the two forms:
+    ///
+    /// - `<SchemaId>` - query a single document
+    /// - `all_<SchemaId>` - query a listing of documents
+    async fn resolve_field(&self, ctx: &Context<'_>) -> ServerResult<Option<Value>> {
+        let field_name = ctx.field().name();
+
+        // Optimistically parse as listing query if field name begins with `all_` (it might still
+        // be a single document query if the schema name itself starts with `all_`).
+        if field_name.starts_with("all_") {
+            if let Ok(schema_id) = field_name.split_at(4).1.parse::<SchemaId>() {
+                // Retrieve the schema to make sure that this is actually a schema and doesn't just
+                // look like one.
+                let schema_provider = ctx.data_unchecked::<SchemaProvider>();
+                if schema_provider.get(&schema_id).await.is_some() {
+                    return self.query_listing(&schema_id, ctx).await;
+                }
+            }
+        }
+
+        // We now know that this is not a listing query. Continue by treating it as a single
+        // document query. Return `Ok(None)` if that doesn't work to signal that other resolvers
+        // should be tried for this field.
+        match field_name.parse::<SchemaId>() {
+            Ok(schema) => self.query_single(&schema, ctx).await,
+            Err(_) => Ok(None),
+        }
+    }
+}
+
+impl DynamicQuery {
+    /// Returns a single document as a GraphQL value.
+    async fn query_single(
+        &self,
+        schema_id: &SchemaId,
```

Comment:
Where does this `SchemaId` come from? The query string right? Sure it's just some `async_graphql` magic...

Reply:
I think its literally just a string from the GraphQL query (like `venue_0020...`) and it gets split at position 4 (after `all_`) and that string is used to parse it (`SchemaId::from_str`)

Reply:
Yes, this method is called from `resolve_field` when the graphql query asks for any field that can be parsed as a schema id. The result of that parsing ends up as the `schema_id` parameter in this method. 

Now that I think about this again I realise that, in contrast to listing queries, there is no check to validate that the schema exists in the schema provider before the view for the query is retrieved, I should add that.